src/compose/service.coffee

```diff
@@ -0,0 +1,709 @@
+			@environment['DOCKER_HOST'] ?= 'unix:///var/run/balena.sock'
+		if checkTruthy(@labels['io.resin.features.supervisor-api'])
+			@_addSupervisorApi(opts)
+		else
+			# We ensure the user hasn't added "supervisor0" to the service's networks
+			delete @networks[constants.supervisorNetworkInterface]
+		if checkTruthy(@labels['io.resin.features.resin-api'])
+			@environment['RESIN_API_KEY'] = opts.deviceApiKey
+
+	extendEnvVars: ({ imageInfo, uuid, appName, name, version, deviceType, osVersion }) =>
+		newEnv =
+			RESIN_APP_ID: @appId.toString()
+			RESIN_APP_NAME: appName
+			RESIN_SERVICE_NAME: @serviceName
+			RESIN_DEVICE_UUID: uuid
+			RESIN_DEVICE_NAME_AT_INIT: name
+			RESIN_DEVICE_TYPE: deviceType
+			RESIN_HOST_OS_VERSION: osVersion
+			RESIN_SUPERVISOR_VERSION: version
+			RESIN_APP_LOCK_PATH: '/tmp/resin/resin-updates.lock'
+			RESIN_SERVICE_KILL_ME_PATH: '/tmp/resin/resin-kill-me'
+			RESIN: '1'
+			USER: 'root'
+		if @environment?
+			_.defaults(newEnv, @environment)
+		_.defaults(newEnv, conversions.envArrayToObject(imageInfo?.Config?.Env ? []))
+		@environment = newEnv
+		return @environment
+
+	extendLabels: (imageInfo) =>
+		@labels = _.clone(@labels)
+		_.defaults(@labels, imageInfo?.Config?.Labels ? {})
+		@labels['io.resin.supervised'] = 'true'
+		@labels['io.resin.app-id'] = @appId.toString()
+		@labels['io.resin.service-id'] = @serviceId.toString()
+		@labels['io.resin.service-name'] = @serviceName
+		return @labels
+
+	extendAndSanitiseExposedPorts: (imageInfo) =>
+		@expose = _.clone(@expose)
+		@expose = _.map(@expose, String)
+		if imageInfo?.Config?.ExposedPorts?
+			for own k, v of imageInfo.Config.ExposedPorts
+				port = k.match(/^([0-9]*)\/tcp$/)?[1]
+				if port? and !_.find(@expose, port)
+					@expose.push(port)
+
+		return @expose
+
+	extendAndSanitiseVolumes: (imageInfo) =>
+		volumes = []
+		for vol in @volumes
+			isBind = _.includes(vol, ':')
+			if isBind
+				[ bindSource, bindDest, mode ] = vol.split(':')
+				if !path.isAbsolute(bindSource)
+					# Rewrite named volumes to namespace by appId
+					volDefinition = "#{@appId}_#{bindSource}:#{bindDest}"
+					if mode?
+						volDefinition += ":#{mode}"
+					volumes.push(volDefinition)
+				else
+					console.log("Ignoring invalid bind mount #{vol}")
+			else
+				volumes.push(vol)
+		volumes = volumes.concat(@defaultBinds())
+		volumes = _.union(_.keys(imageInfo?.Config?.Volumes), volumes)
+		@volumes = volumes
+		return @volumes
+
+	getNamedVolumes: =>
+		defaults = @defaultBinds()
+		validVolumes = _.map @volumes, (vol) ->
+			if _.includes(defaults, vol) or !_.includes(vol, ':')
+				return null
+			bindSource = vol.split(':')[0]
+			if !path.isAbsolute(bindSource)
+				m = bindSource.match(/[0-9]+_(.+)/)
+				return m[1]
+			else
+				return null
+		return _.reject(validVolumes, _.isNil)
+
+	lockPath: =>
+		return updateLock.lockPath(@appId)
+
+	killmePath: =>
+		return killmePath(@appId, @serviceName)
+
+	killmeFullPathOnHost: =>
```

Comment:
For reference there's no difference between fat and thin arrows at the class level - they're both assigned to the prototype in the same way, and I see you've used both here so wanted to check you're aware of that

Reply:
Oh but there is definitely a difference:
```
class A
	constructor: ->
		console.log('created an A')
	slimArrowFn: ->
		console.log(this)
	fatArrowFn: =>
		console.log(this)
```
compiles to:
```
// Generated by CoffeeScript 1.12.2
(function() {
  var A,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  A = (function() {
    function A() {
      this.fatArrowFn = bind(this.fatArrowFn, this);
      console.log('created an A');
    }

    A.prototype.slimArrowFn = function() {
      return console.log(this);
    };

    A.prototype.fatArrowFn = function() {
      return console.log(this);
    };

    return A;

  })();

}).call(this);
```

Notice the `this.fatArrowFn = bind(this.fatArrowFn, this);`. So the fat arrow is safer to use in case we call one of these inside a promise chain e.g.
```
a = new A()
Promise.try(a.fatArrowFn).then(a.slimArrowFn)
```
Here the call to `a.slimArrowFn` will *not* print `a`.

Reply:
(I may have used thin arrows in some places where it'd be better to use a fat one though)

Reply:
Oh ffs, I even compiled it to check and didn't spot that it added it both to the constructor to bind to an instance as well as to the prototype - my bad.